### PR TITLE
[2/N][Attention] Enable masked MHA for sparse MLA prefills

### PR DESCRIPTION
**Pull request by MatthewBonanni**
Depends on https://github.com/vllm-project/flash-attention/pull/155

## Purpose
For pure prefills, there is a threshold of sequence length below which a masked MHA pathway is faster than sparse MQA. The [DeepSeek V3.2 paper mentions](https://arxiv.org/abs/2512.02556) this:

> Note that for short-sequence prefilling, we specially implement a masked MHA mode to
simulate DSA, which can achieve higher efficiency under short-context conditions.

https://github.com/vllm-project/vllm/pull/47327 implemented a dense MHA shortcut for when `seq_len < topk`. This PR extends that optimization by implementing the masked MHA pathway.

## Test Result
mha_vs_mqa_dp_heads128
mha_vs_mqa_tp8_heads16
speedup_heatmap_tp1_tp2_tp4_tp8

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: tests/v1/attention/test_sparse_mla_backends.py [hunk 1]
```diff
@@ -942,13 +945,17 @@ def test_sparse_backend_prefill_correctness(
 
     # Compute dense reference outputs.
     total_query_tokens = sum(query_lens)
-    sparse_indices = torch.zeros(
-        total_query_tokens, topk_tokens, dtype=torch.int32, device=device
+    sparse_indices = torch.full(
+        (total_query_tokens, topk_tokens),
+        -1 if masked_mha else 0,
```
Comment (LucasWilkinson) [inline tests/v1/attention/test_sparse_mla_backends.py]: nit: can you make a note of the -1/0 distinction here? can we not use -1 for both?

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 2]
```diff
@@ -63,7 +65,6 @@ def __init__(
         try:
             self.dcp_world_size = get_dcp_group().world_size
         except AssertionError:
-            # DCP might not be initialized in testing
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py:66]: nit: why remove?

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 3]
```diff
@@ -77,8 +78,6 @@ def __init__(
         )
         workspace_rows = self.chunked_prefill_workspace_size
         if self.dcp_world_size > 1:
-            # DCP gathers each rank's local KV shard into the workspace, so it
-            # needs an extra 1/DCP rows beyond the TP allocation.
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py:81]: nit: why remove?

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 4]
```diff
@@ -231,11 +231,7 @@ def _build_prefill_fields(
         common_attn_metadata: "CommonAttentionMetadata",
         num_decodes: int,
         num_prefills: int,
-    ) -> tuple[
-        torch.Tensor | None,  # prefill_query_start_loc
-        int,  # prefill_max_query_len
-        torch.Tensor | None,  # prefill_query_lens_cpu
-    ]:
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py:242]: nit: why reformat?

File: tests/v1/attention/test_sparse_mla_backends.py (hunk 1 above)
Comment (MatthewBonanni) [reply tests/v1/attention/test_sparse_mla_backends.py]: done in https://github.com/vllm-project/vllm/pull/48770/commits/cfc29572afe14dcafabb319f846691a4cf5a4ca9

File: vllm/model_executor/layers/attention/sparse_mla_attention.py (hunk 2 above)
Comment (MatthewBonanni) [reply vllm/model_executor/layers/attention/sparse_mla_attention.py:66]: fixed in https://github.com/vllm-project/vllm/pull/48770/commits/cfc29572afe14dcafabb319f846691a4cf5a4ca9

File: vllm/model_executor/layers/attention/sparse_mla_attention.py (hunk 3 above)
Comment (MatthewBonanni) [reply vllm/model_executor/layers/attention/sparse_mla_attention.py:81]: fixed in https://github.com/vllm-project/vllm/pull/48770/commits/cfc29572afe14dcafabb319f846691a4cf5a4ca9

File: vllm/model_executor/layers/attention/sparse_mla_attention.py (hunk 4 above)
Comment (MatthewBonanni) [reply vllm/model_executor/layers/attention/sparse_mla_attention.py:242]: fixed in https://github.com/vllm-project/vllm/pull/48770/commits/cfc29572afe14dcafabb319f846691a4cf5a4ca9

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 5]
```diff
@@ -299,20 +407,323 @@ def __init__(
             v_head_dim,
             kv_b_proj,
         )
-
-        # The indexer carries the shared buffer for normal layers and tests;
-        # the explicitly-passed buffer covers backbone skip layers, whose
-        # indexer is not constructed (see deepseek_v2.py).
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py:309]: nit: why remove?

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 6]
```diff
@@ -299,20 +407,323 @@ def __init__(
...
             if indexer is not None
             else topk_indices_buffer
         )
-
         self._use_flashinfer_concat_mla_k = (
             has_flashinfer()
             and which("ninja") is not None
-            and (self.num_heads == 128)
-            and (self.qk_nope_head_dim == 128)
-            and (self.qk_rope_head_dim == 64)
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py:321]: nit: why reformat?

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 7]
```diff
@@ -299,20 +407,323 @@ def __init__(
...
+                assert output_lse is not None
+                if merge_output is None:
+                    merge_output = torch.empty_like(output)
+                    merge_output_lse = torch.empty_like(output_lse)
+                assert merge_output_lse is not None
+                merge_attn_states(
+                    output=merge_output,
+                    output_lse=merge_output_lse,
+                    prefix_output=output,
+                    prefix_lse=output_lse,
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py:714]: can this not be be inplace? i.e.

```
                    output=output,
                    output_lse=output_lse,
                    prefix_output=output,
                    prefix_lse=output_lse,
```

?

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 8]
```diff
@@ -299,20 +407,323 @@ def __init__(
...
+                topk_per_req,
+                q_lens,
+                padded_q_len,
+                prefill_max_seq_len,
+                q.device,
+            )
+            dense_mask = torch.nn.functional.pad(dense_mask, (0, 1))
+            suffix_topk = topk_per_req
+        else:
+            suffix_topk = self._remap_topk_to_ranges(topk_per_req, context_lens, q_lens)
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py]: do we really need both of these branches? i find this very confusing, is there a strong reason we can't stick full to per-chunk masks or gobal masks?

**Review by LucasWilkinson — APPROVED** [review]
Overall looks pretty good to me; left some comments for cleanup/simplification

File: vllm/model_executor/layers/attention/sparse_mla_attention.py (hunk 5 above)
Comment (MatthewBonanni) [reply vllm/model_executor/layers/attention/sparse_mla_attention.py:309]: cleaned up in https://github.com/vllm-project/vllm/pull/48770/commits/7a0c45ad95c52795f36807934e338db111383d2a

File: vllm/model_executor/layers/attention/sparse_mla_attention.py (hunk 6 above)
Comment (MatthewBonanni) [reply vllm/model_executor/layers/attention/sparse_mla_attention.py:321]: cleaned up in https://github.com/vllm-project/vllm/pull/48770/commits/7a0c45ad95c52795f36807934e338db111383d2a

File: vllm/model_executor/layers/attention/sparse_mla_attention.py (hunk 7 above)
Comment (MatthewBonanni) [reply vllm/model_executor/layers/attention/sparse_mla_attention.py:714]: cleaned up in https://github.com/vllm-project/vllm/pull/48770/commits/c68f1a15f67b58128d5f7be3148ab7be53587c1e

File: vllm/model_executor/layers/attention/sparse_mla_attention.py (hunk 8 above)
Comment (MatthewBonanni) [reply vllm/model_executor/layers/attention/sparse_mla_attention.py]: The global mask is a performance win but its size scales with batch_size * query_len * seq_len so we need to establish some reasonable cap, above which we switch to per-chunk masks. I've explicitly defined a `GLOBAL_TOPK_MASK_MAX_BYTES` as 64 MiB.

**Comment by MatthewBonanni** [comment]
Marking as [Do Not Merge] because we still need to land FA once CI is green

**Review by mgoin — APPROVED** [review]
